### ISSUE TITLE
mitmproxy does not handle TLS handshakes for internationalized domain names correctly over Socksv5

### ISSUE THREAD

**Issue by TPXP** [mitmproxy/mitmproxy#3510]
##### Steps to reproduce the problem:

1. `sudo pip install mitmproxy`
2. `mitmproxy --mode socks5 --listen-port 9050`
3. `curl --socks5-hostname 127.0.0.1:9050 -L -k -v https://www.xn--80ak6aa92e.com/`

```
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 9050 (#0)
* found 173 certificates in /etc/ssl/certs/ca-certificates.crt
* found 697 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* gnutls_handshake() failed: An unexpected TLS packet was received.
* Closing connection 0
curl: (35) gnutls_handshake() failed: An unexpected TLS packet was received.
```

##### Any other comments? What have you tried so far?
Versions 0.15, 3.0.4, 2.0.2 and 4.0.4 are also affected by the bug.

Strangely, it works correctly over an HTTP proxy and if I disable remote hostname resolution with SOCKSv5, (use `--socks5` instead of `--socks5-hostname`), it also works correctly.

##### System information
```
Mitmproxy: 4.0.4
Python:    3.7.3
OpenSSL:   OpenSSL 1.1.0h  27 Mar 2018
Platform:  Linux-4.4.0-21-generic-i686-with-Ubuntu-16.04-xenial
```

**Comment by mhils** [comment]
Thanks. Could you provide us with a PCAP?

**Comment by TPXP** [comment]
Hello,

Thank you for your interest in my issue. I'm sorry I was a little busy recently so I didn't have time to make packet captures until now.

I did two packet captures with and without SOCKSv5 hostname resolution and found out that, without the SOCKSv5 hostname resolution (`--socks5` instead of `--socks5-hostname`), everything works correctly. There's probably something wrong in the way the SOCKSv5 proxy answers host-based queries compared to IP-based queries.

[Here are](https://github.com/mitmproxy/mitmproxy/files/3145173/captures.zip) the two packet captures I performed on the loopback interface (between cURL and mitmproxy), with and without SOCKSv5 hostname resolution. I hope it helps!
